File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/transport/internal/RetryableHost.kt
package com.algolia.client.transport.internal

import com.algolia.client.configuration.Host

/**
 * This internal data class represents a host that supports retrying requests.
 *
 * It wraps a Host object and provides additional information about the host's status, such as
 * whether it is up, when it was last updated, and how many times it has been retried.
 *
 * @property host The [Host] object that this RetryableHost is wrapping.
 */
internal data class RetryableHost(
  private val host: Host,
) {

  val url
    get() = host.url
  val callType
    get() = host.callType

  var isUp: Boolean = true
  var lastUpdated: Long = currentTimeMillis()
  var retryCount: Int = 0

  fun reset() {
    lastUpdated = currentTimeMillis()
    isUp = true
    retryCount = 0
  }

  fun hasTimedOut() {
    isUp = true
    lastUpdated = currentTimeMillis()
    retryCount += 1
  }

  fun hasFailed() {
    isUp = false
    lastUpdated = currentTimeMillis()
  }
}

File: clients/algoliasearch-client-dart/packages/client_core/lib/src/transport/retryable_host.dart
import 'package:algolia_client_core/src/config/host.dart';

/// This internal data class represents a host that supports retrying requests.
///
/// It wraps a Host object and provides additional information about the host's status, such as
/// whether it is up, when it was last updated, and how many times it has been retried.
final class RetryableHost {
  final Host host;

  bool get isUp => _isUp;
  bool _isUp;

  DateTime get lastUpdated => _lastUpdated;
  DateTime _lastUpdated;

  int get retryCount => _retryCount;
  int _retryCount;

  RetryableHost(this.host)
      : _isUp = true,
        _lastUpdated = DateTime.now(),
        _retryCount = 0;

  void reset() {
    _isUp = true;
    _lastUpdated = DateTime.now();
    _retryCount = 0;
  }

  void timedOut() {
    _isUp = true;
    _lastUpdated = DateTime.now();
    _retryCount += 1;
  }

  void failed() {
    _isUp = false;
    _lastUpdated = DateTime.now();
  }
}
